core/src/main/java/org/bitcoinj/crypto/bls/BLSHDKeyDerivation.java

```diff
@@ -0,0 +1,126 @@
+/*
+ * Copyright 2013 Matija Mazi.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.bitcoinj.crypto.bls;
+
+import com.google.common.collect.ImmutableList;
+import org.bitcoinj.core.Utils;
+import org.bitcoinj.crypto.ChildNumber;
+import org.bitcoinj.crypto.HDDerivationException;
+import org.dashj.bls.ExtendedPrivateKey;
+import org.dashj.bls.ExtendedPublicKey;
+
+
+import static com.google.common.base.Preconditions.checkArgument;
+
+/**
+ * Implementation of the BLS deterministic wallet child key generation algorithm.
+ */
+public final class BLSHDKeyDerivation {
+    private BLSHDKeyDerivation() { }
+
+    /**
+     * Child derivation may fail (although with extremely low probability); in such case it is re-attempted.
+     * This is the maximum number of re-attempts (to avoid an infinite loop in case of bugs etc.).
+     */
+    public static final int MAX_CHILD_DERIVATION_ATTEMPTS = 100;
```

Comment:
There is overlap between this class and `KeyFactory`.  This class is a copy of `HDKeyDerivation` which also has only static methods.

When the third type of key is added, this class will probably be refactored away and all methods will end up in KeyFactory.